lib/interface_quda.cpp

```diff
@@ -3252,346 +3230,389 @@ void invertQuda(void *hp_x, void *hp_b, QudaInvertParam *param)
+      v_g[0] = milc_fatlink_field;
+      quda::split_field(*collected_milc_fatlink_field, v_g, split_key);
+      v_g[0] = milc_longlink_field;
+      quda::split_field(*collected_milc_longlink_field, v_g, split_key);
     }
 
-    if (getVerbosity() >= QUDA_VERBOSE) {
-      double nin = blas::norm2(in->Component(i));
-      printfQuda("Prepared source %i post mass rescale = %g\n", i, nin);
+    profileInvertMultiSrc.TPSTOP(QUDA_PROFILE_INIT);
+    profileInvertMultiSrc.TPSTART(QUDA_PROFILE_PREAMBLE);
+
+    comm_barrier();
+
+    // Split input fermion field
+    quda::ColorSpinorParam cpu_cs_param_split(*_h_x[0]);
+    for (int d = 0; d < CommKey::n_dim; d++) { cpu_cs_param_split.x[d] *= split_key[d]; }
+    std::vector<quda::ColorSpinorField *> _collect_b(param->num_src_per_sub_partition, nullptr);
+    std::vector<quda::ColorSpinorField *> _collect_x(param->num_src_per_sub_partition, nullptr);
+    for (int n = 0; n < param->num_src_per_sub_partition; n++) {
+      _collect_b[n] = new quda::cpuColorSpinorField(cpu_cs_param_split);
+      _collect_x[n] = new quda::cpuColorSpinorField(cpu_cs_param_split);
+      auto first = _h_b.begin() + n * num_sub_partition;
+      auto last = _h_b.begin() + (n + 1) * num_sub_partition;
+      std::vector<ColorSpinorField *> _v_b(first, last);
+      split_field(*_collect_b[n], _v_b, split_key, pc_type);
     }
-  }
+    comm_barrier();
+
+    push_communicator(split_key);
+    updateR();
+    comm_barrier();
+
+    profileInvertMultiSrc.TPSTOP(QUDA_PROFILE_PREAMBLE);
+    profileInvertMultiSrc.TPSTOP(QUDA_PROFILE_TOTAL);
 
-    // solution_type specifies *what* system is to be solved.
-    // solve_type specifies *how* the system is to be solved.
-    //
-    // We have the following four cases (plus preconditioned variants):
-    //
-    // solution_type    solve_type    Effect
-    // -------------    ----------    ------
-    // MAT              DIRECT        Solve Ax=b
-    // MATDAG_MAT       DIRECT        Solve A^dag y = b, followed by Ax=y
-    // MAT              NORMOP        Solve (A^dag A) x = (A^dag b)
-    // MATDAG_MAT       NORMOP        Solve (A^dag A) x = b
-    // MAT              NORMERR       Solve (A A^dag) y = b, then x = A^dag y
-    //
-    // We generally require that the solution_type and solve_type
-    // preconditioning match.  As an exception, the unpreconditioned MAT
-    // solution_type may be used with any solve_type, including
-    // DIRECT_PC and NORMOP_PC.  In these cases, preparation of the
-    // preconditioned source and reconstruction of the full solution are
-    // taken care of by Dirac::prepare() and Dirac::reconstruct(),
-    // respectively.
-
-    if (pc_solution && !pc_solve) {
-      errorQuda("Preconditioned (PC) solution_type requires a PC solve_type");
+    // Load gauge field after pushing the split communicator so the comm buffers, etc are setup according to
+    // the split topology.
+    printfQuda("Split grid loading gauge field...\n");
```

Comment:
There is a lot of printfQuda in the function. Looks like this is only needed when debugging and should not always be printed by default